src/python/pants/engine/internals/defaults.py

```diff
@@ -0,0 +1,190 @@
+        for tgt, default in defaults.items():
+            if not default:
+                self.defaults.pop(tgt, None)
+            else:
+                self.defaults[tgt] = default
+
+    def _process_defaults(
+        self,
+        defaults: dict[str, dict[str, Any]],
+        targets_defaults: SetDefaultsT,
+        ignore_unknown_fields: bool = False,
+    ):
+        if not isinstance(targets_defaults, dict):
+            raise ValueError(
+                f"Expected dictionary mapping targets to default field values for {self.address} "
+                f"but got: {type(targets_defaults).__name__}."
+            )
+
+        types = self.registered_target_types.aliases_to_types
+        for target, default in targets_defaults.items():
+            if not isinstance(default, dict):
+                raise ValueError(
+                    f"Invalid default field values in {self.address} for target type {target}, "
+                    f"must be an `dict` but was {default!r} with type `{type(default).__name__}`."
+                )
+
+            targets: Iterable[str]
+            targets = target if isinstance(target, tuple) else (target,)
+            for target_alias in map(str, targets):
+                if target_alias in types:
+                    target_type = types[target_alias]
+                else:
+                    raise ValueError(f"Unrecognized target type {target_alias} in {self.address}.")
+
+                # Copy default dict if we may mutate it.
+                raw_values = dict(default) if ignore_unknown_fields else default
+
+                # Validate that field exists on target
+                target_fields = target_type.class_field_types(self.union_membership)
+                valid_field_aliases = set()
+
+                # TODO: this valid aliases calculation is done every time a target is instantiated
+                # as well. But it should be enough to do once, and re-use as it doesn't change
+                # during a run.
+                for fld in target_fields:
+                    valid_field_aliases.add(fld.alias)
+                    if fld.deprecated_alias is not None:
+                        valid_field_aliases.add(fld.deprecated_alias)
+
+                for field_alias in default.keys():
+                    if field_alias not in valid_field_aliases:
+                        if ignore_unknown_fields:
+                            del raw_values[field_alias]
+                        else:
+                            raise InvalidFieldException(
+                                f"Unrecognized field `{field_alias}` for target {target_type.alias}. "
+                                f"Valid fields are: {', '.join(sorted(valid_field_aliases))}.",
+                            )
+                # TODO: moved fields for TargetGenerators ?  See: `Target._calculate_field_values()`.
+                # TODO: support parametrization ?
```

Comment:
I think that parametrization should "just work", because the `parametrize` object is preserved, and not expanded until the target is actually constructed.

Reply:
I think the issue here would be that we do eager validation of the default values when freezing them.. so we would be passing a Parametrize object to the field.compute_value.

Reply:
Even so, we can avoid this validation, but get stuck on sticking the `Parametrize` instance in a `FrozenDict` does not work well: 
```
E           TypeError: Even though you are using a `FrozenDict`, the underlying values are not hashable. Please use hashable (and preferably immutable) types for the underlying values, e.g. use tuples instead of lists and use FrozenOrderedSet instead of set().
E           
E           Original error message: unhashable type: 'dict'
E           
E           Value: FrozenDict({'description': parametrize(a=desc A, b=desc B})
```